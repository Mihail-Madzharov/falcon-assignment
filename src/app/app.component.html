<swal
  #sweetAlert
  (confirm)="onSweetAlertConfirm()"
  (cancel)="onSweetAlertCancel()"
>
</swal>
<router-outlet></router-outlet>
