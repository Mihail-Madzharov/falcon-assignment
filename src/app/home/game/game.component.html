<main class="mdl-layout__content flc-layout--center-middle">
  <div class="page-content flc-layout--center">
    <div class="mdl-card mdl-shadow--2dp flc-card-main">
      <app-game-board
        [canUserPlay]="
          checkIfPlayerCanPlay(
            gameStarted$ | async,
            currentUserId$ | async,
            lastPlayingPlayerId$ | async,
            winnerId$ | async
          )
        "
        [gameBoard]="gameBoard$ | async"
        (selectCell)="onCellSelect($event)"
      ></app-game-board>
    </div>
    <div class="mdl-card mdl-shadow--2dp flc-card-aside">
      <app-players
        [lastPlayingPlayerId]="lastPlayingPlayerId$ | async"
        (startGame)="onStartGameClickHandler()"
        [currentUserId]="currentUserId$ | async"
        [secondUserId]="secondUserId$ | async"
        [gameStarted]="gameStarted$ | async"
        [winnerId]="winnerId$ | async"
        (joinGame)="onJoinGame()"
        (newGame)="onNewGameClickHandler()"
        (leaveGame)="onLeaveGameClickHandler()"
      ></app-players>
    </div>
  </div>
</main>
