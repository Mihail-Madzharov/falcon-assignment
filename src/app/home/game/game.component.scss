@import "../../../variables";

@keyframes winner-jump {
  from {
    transform: translateY(0) rotateY(0);
  }
  to {
    transform: translateY(-7px) rotateY(20deg);
  }
}
@keyframes drop-in {
  from {
    transform: translateY(-$board-height);
  }
  to {
    transform: translateY(0);
  }
}
@keyframes thinking {
  from {
    transform: rotateZ(-45deg) translate(-12px, 45px);
  }
  to {
    transform: rotateZ(-50deg) translate(-12px, 45px);
  }
}

.flc-game-avatar-group {
  display: flex;
  align-items: center;
  margin-bottom: 10px;
}
.flc-game-avatar {
  display: inline-block;
  height: $avatar-size;
  width: $avatar-size;
  border: 1px solid #eee;
  border-radius: 100%;
  background-size: contain;
  background-repeat: no-repeat;
  background-color: #fff;
  margin: $avatar-margin;
  background-position-y: 2px;
  overflow: hidden;
  box-sizing: border-box;
  &:after {
    content: "";
    background-image: url("../../../assets/hand.svg");
    background-repeat: none;
    height: 30px;
    width: 30px;
    background-size: contain;
    display: none;
  }
  &.man {
    background-image: url("../../../assets/man.svg");
  }
  &.woman {
    background-image: url("../../../assets/woman.svg");
  }
}

.player-one {
  background-color: $player-one-color;
  border: 1px solid green;
}
.player-two {
  background-color: $player-two-color;
  border: 1px solid green;
}
.winner {
  animation: winner-jump 0.3s cubic-bezier(0.17, 0.67, 0.83, 0.67) infinite
    alternate-reverse !important;
  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.2);
}
.their-turn {
  border-width: 2px;
  &:after {
    display: inline-block;
    animation: thinking 2s linear infinite alternate-reverse !important;
  }
}

.flc-card {
  &-main {
    margin: 10px;
    width: 100%;
  }
  &-aside {
    margin: 10px;
    padding: 10px;
    box-sizing: border-box;
    width: 390px;
  }
}
.flc-layout {
  &-bg {
    background-color: #333;
    background-image: repeating-linear-gradient(
        120deg,
        rgba(255, 255, 255, 0.1),
        rgba(255, 255, 255, 0.1) 1px,
        transparent 1px,
        transparent 60px
      ),
      repeating-linear-gradient(
        60deg,
        rgba(255, 255, 255, 0.1),
        rgba(255, 255, 255, 0.1) 1px,
        transparent 1px,
        transparent 60px
      ),
      linear-gradient(
        60deg,
        rgba(0, 0, 0, 0.1) 25%,
        transparent 25%,
        transparent 75%,
        rgba(0, 0, 0, 0.1) 75%,
        rgba(0, 0, 0, 0.1)
      ),
      linear-gradient(
        120deg,
        rgba(0, 0, 0, 0.1) 25%,
        transparent 25%,
        transparent 75%,
        rgba(0, 0, 0, 0.1) 75%,
        rgba(0, 0, 0, 0.1)
      );
    background-size: 70px 120px;
  }

  &--center {
    display: flex;
    align-items: flex-start;
    justify-content: center;
  }
  &--center-middle {
    display: flex;
    align-items: center;
    justify-content: center;
  }
}

.flc-card {
  &-main {
    margin: 10px;
    width: 100%;
  }
  &-aside {
    margin: 10px;
    padding: 10px;
    box-sizing: border-box;
    width: 390px;
  }
}
