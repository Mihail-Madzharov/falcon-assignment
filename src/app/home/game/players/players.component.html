<app-avatar
  [isCurrentPlayer]="currentUserId !== lastPlayingPlayerId"
  [theirTurn]="
    PLAYER_TWO_ID === lastPlayingPlayerId &&
    lastPlayingPlayerId == currentUserId
  "
  [label]="'Player one'"
  [subtitle]="calculateCurrentUserSubLabel()"
  [gameStarted]="currentUserId === PLAYER_ONE_ID"
  [winner]="winnerId === PLAYER_ONE_ID"
></app-avatar>

<app-avatar
  [isCurrentPlayer]="currentUserId !== lastPlayingPlayerId"
  [theirTurn]="
    PLAYER_ONE_ID === lastPlayingPlayerId &&
    lastPlayingPlayerId == currentUserId
  "
  [winner]="winnerId === PLAYER_TWO_ID"
  [label]="'Player two'"
  [subtitle]="calculateSecondUserSubLabel()"
  [gameStarted]="currentUserId === PLAYER_TWO_ID"
></app-avatar>

<div class="mdl-card__actions mdl-card--border">
  <button
    class="mdl-button mdl-js-button mdl-button--raised mdl-button--accent mdl-js-ripple-effect"
    (click)="onStartGameHandler()"
    [disabled]="
      (!gameStarted && currentUserId !== NO_PLAYER_SELECTED_ID) ||
      secondUserId !== NO_PLAYER_SELECTED_ID
    "
  >
    Start game
  </button>
  <button
    *ngIf="secondUserId !== -1 && !gameStarted"
    class="mdl-button mdl-js-button mdl-button--accent mdl-js-ripple-effect"
    (click)="onJoinGame()"
  >
    Join game
  </button>
</div>
