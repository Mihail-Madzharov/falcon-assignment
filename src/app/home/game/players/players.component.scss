@import "../../../../variables";

@keyframes winner-jump {
  from {
    transform: translateY(0) rotateY(0);
  }
  to {
    transform: translateY(-7px) rotateY(20deg);
  }
}

@keyframes thinking {
  from {
    transform: rotateZ(-45deg) translate(-12px, 45px);
  }
  to {
    transform: rotateZ(-50deg) translate(-12px, 45px);
  }
}

.flc-game-avatar-group {
  display: flex;
  align-items: center;
  margin-bottom: 10px;
}
.flc-game-avatar {
  display: inline-block;
  height: $avatar-size;
  width: $avatar-size;
  border: 1px solid #eee;
  border-radius: 100%;
  background-size: contain;
  background-repeat: no-repeat;
  background-color: #fff;
  margin: $avatar-margin;
  background-position-y: 2px;
  overflow: hidden;
  box-sizing: border-box;
  &:after {
    content: "";
    background-image: url("../../../../assets/hand.svg");
    background-repeat: none;
    height: 30px;
    width: 30px;
    background-size: contain;
    display: none;
  }
  &.man {
    background-image: url("../../../../assets/man.svg");
  }
  &.woman {
    background-image: url("../../../../assets/woman.svg");
  }
}

.player-one {
  background-color: $player-one-color;
  border: 1px solid green;
}
.player-two {
  background-color: $player-two-color;
  border: 1px solid green;
}
.winner {
  animation: winner-jump 0.3s cubic-bezier(0.17, 0.67, 0.83, 0.67) infinite
    alternate-reverse !important;
  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.2);
}
.their-turn {
  border-width: 2px;
  &:after {
    display: inline-block;
    animation: thinking 2s linear infinite alternate-reverse !important;
  }
}
