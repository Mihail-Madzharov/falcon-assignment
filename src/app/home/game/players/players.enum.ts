export enum PlayersEnum {
  PlayerOne = 1,
  PlayerTwo = 2
}
