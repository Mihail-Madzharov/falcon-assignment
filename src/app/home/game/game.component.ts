import { Component, OnInit, Inject, OnDestroy } from "@angular/core";
import { Observable, Subject } from "rxjs";
import { Dispatcher } from "src/app/shared/types";
import { tap, takeUntil } from "rxjs/operators";

import { Matrix } from "src/app/lib/game-utilities/matrix";
import { DispatcherToken } from "src/app/app.tokens";
import {
  GameBoardToken,
  CurrentUserIdToken,
  SecondUserIdToken,
  GameStartedToken,
  LastPlayingPlayerId,
  WinnerIdToken
} from "./store/game.token";
import {
  StartGameAction,
  SelectCellAction,
  UpdateSecondUserId,
  UpdateCurrentUserIdAction,
  ToggleGameStartAction,
  UpdateLastPlayingPlayer,
  ResetGameStateAction,
  StartNewGame
} from "./store/game.actions";
import { BoardCell } from "./game-board/board-cell";
import { PlayersEnum } from "./players/players.enum";
import { WebSocketService } from "../web-socket.service";

@Component({
  selector: "app-game",
  templateUrl: "./game.component.html",
  styleUrls: ["./game.component.scss"]
})
export class GameComponent implements OnInit, OnDestroy {
  private destroy$ = new Subject();

  constructor(
    @Inject(GameBoardToken)
    public gameBoard$: Observable<Matrix>,
    @Inject(DispatcherToken)
    private dispatcher: Dispatcher,
    @Inject(CurrentUserIdToken)
    public currentUserId$: Observable<number>,
    @Inject(SecondUserIdToken)
    public secondUserId$: Observable<number>,
    private webSockets: WebSocketService,
    @Inject(GameStartedToken)
    public gameStarted$: Observable<boolean>,
    @Inject(LastPlayingPlayerId)
    public lastPlayingPlayerId$: Observable<number>,
    @Inject(WinnerIdToken)
    public winnerId$: Observable<number>
  ) {}

  ngOnInit(): void {
    this.webSockets.createWebSocketConnection();

    this.webSockets
      .getDownstream()
      .pipe(
        takeUntil(this.destroy$),
        tap(event => {
          if (event.data.message.globalType != null) {
            this.dispatcher(event.data.message);
          }
        })
      )
      .subscribe();
  }

  ngOnDestroy(): void {
    this.destroy$.next();
    this.destroy$.complete();
  }

  onStartGameClickHandler() {
    this.webSockets.send(new StartGameAction());
    this.webSockets.send(new UpdateSecondUserId(PlayersEnum.PlayerOne));

    this.dispatcher(new StartGameAction());
    this.dispatcher(new UpdateCurrentUserIdAction(PlayersEnum.PlayerOne));
  }

  onCellSelect(cell: BoardCell) {
    this.dispatcher(new SelectCellAction(cell));
  }

  onJoinGame() {
    this.webSockets.send(new ToggleGameStartAction(true));
    this.webSockets.send(new UpdateSecondUserId(PlayersEnum.PlayerTwo));
    this.webSockets.send(new UpdateLastPlayingPlayer(PlayersEnum.PlayerTwo));

    this.dispatcher(new ToggleGameStartAction(true));
    this.dispatcher(new UpdateCurrentUserIdAction(PlayersEnum.PlayerTwo));
    this.dispatcher(new UpdateLastPlayingPlayer(PlayersEnum.PlayerTwo));
  }

  checkIfPlayerCanPlay(
    gameStarted: boolean,
    currentUserId: number,
    lastPlayingPlayerId: number
  ) {
    return gameStarted && currentUserId !== lastPlayingPlayerId;
  }

  onNewGameClickHandler() {
    this.dispatcher(new StartNewGame());
    this.webSockets.send(new StartNewGame());
  }

  onLeaveGameClickHandler() {
    this.dispatcher(new ResetGameStateAction());
    this.webSockets.send(new ResetGameStateAction());
  }
}
