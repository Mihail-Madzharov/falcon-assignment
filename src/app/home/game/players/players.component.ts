import { Component, Input, Output, EventEmitter } from "@angular/core";
import { PlayersEnum } from "./players.enum";

@Component({
  selector: "app-players",
  templateUrl: "./players.component.html",
  styleUrls: ["./players.component.scss"]
})
export class PlayersComponent {
  public PLAYER_ONE_ID: number;
  public PLAYER_TWO_ID: number;
  public NO_PLAYER_SELECTED_ID: number;
  @Input()
  gameStarted: boolean;

  @Output()
  startGame = new EventEmitter<boolean>();

  @Output()
  newGame = new EventEmitter();

  @Output()
  leaveGame = new EventEmitter();

  @Output()
  joinGame = new EventEmitter();

  @Input()
  secondUserId: number;

  @Input()
  currentUserId: number;

  @Input()
  lastPlayingPlayerId: number;

  @Input()
  winnerId: number;

  constructor() {
    this.PLAYER_ONE_ID = PlayersEnum.PlayerOne;
    this.PLAYER_TWO_ID = PlayersEnum.PlayerTwo;
    this.NO_PLAYER_SELECTED_ID = PlayersEnum.NoPlayerSelectedId;
  }

  onStartGameHandler() {
    this.startGame.emit();
  }

  onJoinGame() {
    this.joinGame.emit();
  }

  calculateSecondUserSubLabel() {
    if (
      this.secondUserId === this.NO_PLAYER_SELECTED_ID &&
      this.currentUserId === this.PLAYER_ONE_ID
    ) {
      return "(waiting to join)";
    }

    if (this.currentUserId === this.PLAYER_TWO_ID) {
      return "(you)";
    }
  }

  calculateCurrentUserSubLabel() {
    if (this.secondUserId !== this.NO_PLAYER_SELECTED_ID && !this.gameStarted) {
      return "(waiting for you to join)";
    }

    if (this.currentUserId === this.PLAYER_ONE_ID) {
      return "(you)";
    }
  }

  onNewGameClickHandler() {
    this.newGame.emit();
  }

  onLeaveClickHandler() {
    this.leaveGame.emit();
  }
}
