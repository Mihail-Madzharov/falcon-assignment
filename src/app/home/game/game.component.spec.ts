import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { CUSTOM_ELEMENTS_SCHEMA, Pipe, PipeTransform } from "@angular/core";
import { of, Subject } from "rxjs";

import { GameComponent } from "./game.component";
import {
  GameBoardToken,
  CurrentUserIdToken,
  SecondUserIdToken,
  GameStartedToken,
  LastPlayingPlayerId,
  WinnerIdToken
} from "./store/game.token";
import { DispatcherToken } from "src/app/app.tokens";
import { WebSocketService } from "../web-socket.service";
import { GameActionTypes } from "./store/game.actions";
import { PlayersEnum } from "./players/players.enum";
@Pipe({
  name: "async"
})
export class AsyncPipeMock implements PipeTransform {
  transform(value: any, ...args: any[]) {
    return [value];
  }
}

describe("GameComponent", () => {
  let component: GameComponent;
  let fixture: ComponentFixture<GameComponent>;
  beforeEach(async(() => {
    const dispatcher = jest.fn();

    TestBed.configureTestingModule({
      declarations: [GameComponent, AsyncPipeMock],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      providers: [
        { provide: GameBoardToken, useValue: {} },
        {
          provide: DispatcherToken,
          useValue: dispatcher
        },
        {
          provide: CurrentUserIdToken,
          useValue: () => {}
        },
        {
          provide: SecondUserIdToken,
          useValue: () => {}
        },
        {
          provide: GameStartedToken,
          useValue: () => {}
        },
        {
          provide: LastPlayingPlayerId,
          useValue: () => {}
        },
        {
          provide: WinnerIdToken,
          useValue: () => {}
        },
        WebSocketService
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(GameComponent);
    component = fixture.componentInstance;
  });

  it("should create", () => {
    expect(component).toBeTruthy();
  });

  it("should match snapshot", () => {
    expect(fixture).toMatchSnapshot();
  });

  describe("", () => {
    it(`should dispatch action to start the game and
  send actions through the websocket to update the other user state`, () => {
      // Arrange
      const dispatcherActionContainers = [];
      const mockDispatcher = action => {
        dispatcherActionContainers.push(action);
      };

      const webSocketsActionContainers = [];
      const mockWebSocketService = {
        send: action => {
          webSocketsActionContainers.push(action);
        }
      };
      const comp = setup()
        .default()
        .mockDispatcher(mockDispatcher)
        .mockWebSocketService(mockWebSocketService)
        .build();

      // Act
      comp.onStartGameClickHandler();

      // webSocket Assert
      expect(webSocketsActionContainers[0].type).toBe(
        GameActionTypes.StartGame
      );
      expect(webSocketsActionContainers[1].type).toBe(
        GameActionTypes.UpdateSecondUserId
      );
      expect(webSocketsActionContainers[1].payload).toBe(PlayersEnum.PlayerOne);

      // Dispatcher Assert
      expect(dispatcherActionContainers[0].type).toBe(
        GameActionTypes.StartGame
      );
      expect(dispatcherActionContainers[1].type).toBe(
        GameActionTypes.UpdateCurrentUserId
      );
      expect(dispatcherActionContainers[1].payload).toBe(PlayersEnum.PlayerOne);
    });
  });

  describe("onCellSelect", () => {
    it(`should dispatch action for selecting a cell and call send function of the webSocketService to update the selected cell`, () => {
      // Arrange
      const dispatcherActionContainers = [];
      const mockDispatcher = action => {
        dispatcherActionContainers.push(action);
      };

      const comp = setup()
        .default()
        .mockDispatcher(mockDispatcher)
        .build();

      // Act
      comp.onCellSelect({ col: 1, row: 1 });

      // Dispatcher Assert
      expect(dispatcherActionContainers[0].type).toBe(
        GameActionTypes.SelectCell
      );
      expect(dispatcherActionContainers[0].payload).toEqual({ col: 1, row: 1 });
    });
  });

  describe("onJoinGame", () => {
    it(`should dispatch action for joining a game and call send function of the webSocketService to notify the other user`, () => {
      // Arrange
      const dispatcherActionContainers = [];
      const mockDispatcher = action => {
        dispatcherActionContainers.push(action);
      };

      const webSocketsActionContainers = [];
      const mockWebSocketService = {
        send: action => {
          webSocketsActionContainers.push(action);
        }
      };
      const comp = setup()
        .default()
        .mockDispatcher(mockDispatcher)
        .mockWebSocketService(mockWebSocketService)
        .build();

      // Act
      comp.onJoinGame();

      // webSocket Assert
      expect(webSocketsActionContainers[0].type).toBe(
        GameActionTypes.ToggleStartGame
      );
      expect(webSocketsActionContainers[0].payload).toBeTruthy();

      expect(webSocketsActionContainers[1].type).toBe(
        GameActionTypes.UpdateSecondUserId
      );
      expect(webSocketsActionContainers[1].payload).toEqual(
        PlayersEnum.PlayerTwo
      );

      expect(webSocketsActionContainers[2].type).toBe(
        GameActionTypes.UpdateLastPlayingPlayer
      );
      expect(webSocketsActionContainers[2].payload).toEqual(
        PlayersEnum.PlayerTwo
      );

      // Dispatcher Assert
      expect(dispatcherActionContainers[0].type).toBe(
        GameActionTypes.ToggleStartGame
      );
      expect(dispatcherActionContainers[0].payload).toBeTruthy();
      expect(dispatcherActionContainers[1].type).toBe(
        GameActionTypes.UpdateCurrentUserId
      );
      expect(dispatcherActionContainers[1].payload).toEqual(
        PlayersEnum.PlayerTwo
      );
      expect(dispatcherActionContainers[2].type).toBe(
        GameActionTypes.UpdateLastPlayingPlayer
      );
      expect(dispatcherActionContainers[2].payload).toEqual(
        PlayersEnum.PlayerTwo
      );
    });
  });

  describe("ngOnInit", () => {
    it("should call createWebsocketConnection on init", () => {
      // Arrange
      const createConnectionSpy = jest.fn();
      const mockWebSocketService = {
        createWebSocketConnection: createConnectionSpy,
        getDownstream: action => of()
      };

      const comp = setup()
        .default()
        .mockWebSocketService(mockWebSocketService)
        .build();
      // Act

      comp.ngOnInit();

      // Assert
      expect(createConnectionSpy).toHaveBeenCalled();
    });

    it("should dispatch action coming from the downStreamData if it type is globalType", () => {
      // Arrange
      const dispatcherActionContainers = [];
      const mockDispatcher = action => {
        dispatcherActionContainers.push(action);
      };

      const mockDownStreamData = new Subject();
      const mockWebSocketService = {
        createWebSocketConnection: () => {},
        getDownstream: action => mockDownStreamData
      };

      const comp = setup()
        .default()
        .mockDispatcher(mockDispatcher)
        .mockWebSocketService(mockWebSocketService)
        .build();
      comp.ngOnInit();
      // Act
      mockDownStreamData.next({
        data: {
          message: {
            globalType: "Mock global type",
            type: GameActionTypes.StartGame
          }
        }
      });
      // Assert
      expect(dispatcherActionContainers[0].type).toBe(
        GameActionTypes.StartGame
      );
    });
  });

  describe("checkIfPlayerCanPlay", () => {
    it("should return false if game started is false and the currentUserId is different than the lastPlayingPlayerId", () => {
      // Act
      const result = component.checkIfPlayerCanPlay(false, 1, 2);
      // Assert
      expect(result).toBeFalsy();
    });

    it("should return true if game started is true and the currentUserId is different than the lastPlayingPlayerId", () => {
      // Arrange
      // Act
      const result = component.checkIfPlayerCanPlay(true, 1, 1);
      // Assert
      expect(result).toBeFalsy();
    });
  });

  describe("onNewGameClickHandler", () => {
    it("should dispatch StartNewGame and send StartNewGame", () => {
      // Arrange
      const dispatcherActionContainers = [];
      const mockDispatcher = action => {
        dispatcherActionContainers.push(action);
      };
      const webSocketsActionContainers = [];
      const mockWebSocketService = {
        send: action => {
          webSocketsActionContainers.push(action);
        }
      };

      const comp = setup()
        .default()
        .mockDispatcher(mockDispatcher)
        .mockWebSocketService(mockWebSocketService)
        .build();

      // Act
      comp.onNewGameClickHandler();

      // Dispatcher Assert
      expect(dispatcherActionContainers[0].type).toBe(
        GameActionTypes.StartNewGame
      );
      expect(webSocketsActionContainers[0].type).toBe(
        GameActionTypes.StartNewGame
      );
    });
  });

  describe("onLeaveGameClickHandler", () => {
    it("should dispatch ResetGameStateAction and send ResetGameStateAction", () => {
      // Arrange
      const dispatcherActionContainers = [];
      const mockDispatcher = action => {
        dispatcherActionContainers.push(action);
      };
      const webSocketsActionContainers = [];
      const mockWebSocketService = {
        send: action => {
          webSocketsActionContainers.push(action);
        }
      };

      const comp = setup()
        .default()
        .mockDispatcher(mockDispatcher)
        .mockWebSocketService(mockWebSocketService)
        .build();

      // Act
      comp.onLeaveGameClickHandler();

      // Dispatcher Assert
      expect(dispatcherActionContainers[0].type).toBe(
        GameActionTypes.ResetGameState
      );
      expect(webSocketsActionContainers[0].type).toBe(
        GameActionTypes.ResetGameState
      );
    });
  });
  function setup() {
    let gameBoard;
    let currentUserId$;
    let secondUserId$;
    let gameStarted$;
    let lastPlayingPlayerId$;
    let dispatcher;
    let webSockets;
    let winnerId$;
    const builder = {
      gameBoard,
      currentUserId$,
      secondUserId$,
      winnerId$,
      build: () => {
        return new GameComponent(
          gameBoard,
          dispatcher,
          currentUserId$,
          secondUserId$,
          webSockets,
          gameStarted$,
          lastPlayingPlayerId$,
          winnerId$
        );
      },
      default: () => {
        gameBoard = of([]);
        currentUserId$ = of(1);
        secondUserId$ = of(1);
        gameStarted$ = of(false);
        lastPlayingPlayerId$ = of(0);
        dispatcher = () => {};
        webSockets = {} as any;
        winnerId$ = of(1);
        return builder;
      },
      mockDispatcher: dispatcherMock => {
        dispatcher = dispatcherMock;
        return builder;
      },
      mockWebSocketService: (mockWebSocket: any) => {
        webSockets = mockWebSocket;
        return builder;
      }
    };
    return builder;
  }
});
