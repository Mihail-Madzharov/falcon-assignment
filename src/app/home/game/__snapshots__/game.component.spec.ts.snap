// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameComponent should match snapshot 1`] = `
<app-game
  currentUserId$={[Function Function]}
  destroy$={[Function Subject]}
  dispatcher={[Function Function]}
  gameBoard$={[Function Object]}
  gameStarted$={[Function Function]}
  lastPlayingPlayerId$={[Function Function]}
  secondUserId$={[Function Function]}
  webSockets={[Function WebSocketService]}
>
  <main
    class="mdl-layout__content flc-layout--center-middle"
  >
    <div
      class="page-content flc-layout--center"
    >
      <div
        class="mdl-card mdl-shadow--2dp flc-card-main"
      >
        <app-game-board />
      </div>
      <div
        class="mdl-card mdl-shadow--2dp flc-card-aside"
      >
        <app-players />
      </div>
    </div>
  </main>
</app-game>
`;
